### PR TITLE
UI: Reduce settings dialog default height

### LINKED ISSUE #5168: Settings Window height is more than my screen height
### Operating System Info

Windows 10

### Other OS

_No response_

### OBS Studio Version

27.0.1

### OBS Studio Version (Other)

_No response_

### OBS Studio Log URL

https://obsproject.com/logs/MlsqgVTdI6HoGoT9

### OBS Studio Crash Log URL

_No response_

### Expected Behavior

When I click on Settings button on the Controls panel, settings window should open with its height <= screen/display height. So that "OK", "Cancel" and & "Apply" buttons are visible on the screen/display.

### Current Behavior

When I click on Settings button on the Controls panel, settings window is opening with its height > screen/display height. Due to this, "OK", "Cancel" and & "Apply" buttons are going outside of screen/display area.

https://user-images.githubusercontent.com/43258549/130329102-f83f54b2-7ce3-44c1-90cd-fd84ab894aa3.mp4

### Steps to Reproduce

1. Open OBS
2. Click on Settings button on Controls panel
...

### Anything else we should know?

Please find the video attached in the Current Behavior section.

### PR DESCRIPTION
**Pull request by tytan652**
### Description

Reduce the size of the settings window if greater than the current display's available geometry.

### Motivation and Context

Fixes #5168

This change should avoid this to happen again on 720p:

[Example]()
OK, Cancel and Apply are not accessible without having to resize the window.
### How Has This Been Tested?

- Set manually the main window size at something 720p-like.
- Open the settings window.
- The settings window is smaller in height than the main window.

### Types of changes

- Tweak (non-breaking change to improve existing functionality)

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

**Comment by WizardCM** [comment]
It annoys me that Qt in this case is ignoring the monitor's height and treating the default height more like a minimum height. This should be automatic - "if default window size doesn't fit, resize to be smaller, otherwise stay default". Maybe there's a way for us to force it? Maybe we're setting some default heights of internal widgets somewhere causing it to stretch to its default regardless?

**Comment by tytan652** [comment]
Made the fix in code rather than in UI file.

**Comment by WizardCM** [comment]
I wonder if we could use QScreen::availableGeometry to determine the safest maximum size for a window on the current display. Might provide a slightly better experience than basing it on the height of the main window. However, will need to be tested on Wayland.

**Review by Warchamp7 — APPROVED** [review]
On one hand, basing this off the size of the main window feels janky.

On the other hand, if someone has made the main window that small despite having more screen real estate, it's not the worst thing to respect that decision in the settings window too.

**Comment by RytoEX** [comment]
> On one hand, basing this off the size of the main window feels janky.
> 
> On the other hand, if someone has made the main window that small despite having more screen real estate, it's not the worst thing to respect that decision in the settings window too.

I think it would be better to use the screen dimensions instead of the main window.

**Comment by tytan652** [comment]
> On the other hand, if someone has made the main window that small despite having more screen real estate, it's not the worst thing to respect that decision in the settings window too.

If the asked height is too small for the windows, it will normally not respect the asked size change.

> I think it would be better to use the screen dimensions instead of the main window.

We can't because of the Windows 10/11 taskbar which is the main cause of this PR.

**Comment by RytoEX** [comment]
> > On the other hand, if someone has made the main window that small despite having more screen real estate, it's not the worst thing to respect that decision in the settings window too.
> 
> If the asked height is too small for the windows, it will normally not respect the asked size change.
> 
> > I think it would be better to use the screen dimensions instead of the main window.
> 
> We can't because of the Windows 10/11 taskbar which is the main cause of this PR.

You can just use 85/90% of the screen dimensions to account for the taskbar's minimum size, which should be sufficient in the vast majority of cases, if not all default cases.

**Comment by PatTheMav** [comment]
@RytoEX While not perfect this seems reasonable to me? Got any objections to having this merged?

File: UI/window-basic-settings.cpp [hunk 1]
```diff
@@ -921,6 +921,13 @@ OBSBasicSettings::OBSBasicSettings(QWidget *parent)
 
 	UpdateAudioWarnings();
 	UpdateAdvNetworkGroup();
+
+	/* Reduce the height of the window if too tall compared to the screen
+	 * size (e.g. 720p) with potential desktop decoration (e.g. taskbar) */
+	int max_height = round(
+		QGuiApplication::primaryScreen()->size().height() * 95 / 100);
```
Comment (Warchamp7) [inline UI/window-basic-settings.cpp:928]: Does primaryScreen in this case get the primary display or the display that OBS is actually running on?

Comment (RytoEX) [reply UI/window-basic-settings.cpp:928]: The primary display.

Comment (Warchamp7) [reply UI/window-basic-settings.cpp:928]: @tytan652 We should base this off the dimensions of the screen that the settings dialog is actually opening on

Comment (tytan652) [reply UI/window-basic-settings.cpp]: Made the change but it needs to be tested on Windows.

**Comment by RytoEX** [comment]
Off-thread review yielded that the constructor is too early for this to occur, as the widget does not yet have its screen updated. Overloading showEvent and handling this there is late enough for the widget's screen to be correct. @Warchamp7 is driving this.

**Comment by RytoEX** [comment]
Rebased. Will clean up commits and implement some changes from off-thread review.

**Review by RytoEX — APPROVED** [review]
Tested the adjusted PR changes locally on Windows 11 with a 1280x720 display. The Settings window is resized to _exactly_ fit in the usable screen space.

**Pull request merged**